src/OpenTelemetry.Contrib.Instrumentation.GrpcCore/ServerTracingInterceptor.cs

```diff
@@ -0,0 +1,214 @@
+
+                var responseStreamProxy = new ServerStreamWriterProxy<TResponse>(
+                    responseStream,
+                    rpcScope.RecordResponse);
+
+                await continuation(requestStreamReaderProxy, responseStreamProxy, context).ConfigureAwait(false);
+                rpcScope.Complete();
+            }
+            catch (Exception e)
+            {
+                rpcScope.CompleteWithException(e);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// A class to help track the lifetime of a service-side RPC.
+        /// </summary>
+        /// <typeparam name="TRequest">The type of the request.</typeparam>
+        /// <typeparam name="TResponse">The type of the response.</typeparam>
+        private class ServerRpcScope<TRequest, TResponse> : RpcScope<TRequest, TResponse>
+            where TRequest : class
+            where TResponse : class
+        {
+            /// <summary>
+            /// The metadata setter action.
+            /// </summary>
+            private static readonly Func<Metadata, string, IEnumerable<string>> MetadataGetter = (metadata, key) =>
+            {
+                for (var i = 0; i < metadata.Count; i++)
+                {
+                    var entry = metadata[i];
+                    if (entry.Key.Equals(key))
+                    {
+                        return new string[1] { entry.Value };
+                    }
+                }
+
+                return Enumerable.Empty<string>();
+            };
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="ServerRpcScope{TRequest, TResponse}"/> class.
+            /// </summary>
+            /// <param name="context">The context.</param>
+            /// <param name="options">The options.</param>
+            public ServerRpcScope(ServerCallContext context, ServerTracingInterceptorOptions options)
+                : base(context.Method, options.RecordMessageEvents)
+            {
+                if (!GrpcCoreInstrumentation.ActivitySource.HasListeners())
+                {
+                    return;
+                }
+
+                var currentContext = Activity.Current?.Context;
+
+                // Extract the SpanContext, if any from the headers
+                var metadata = context.RequestHeaders;
+                if (metadata != null)
+                {
+                    var propagationContext = options.Propagator.Extract(new PropagationContext(currentContext ?? default, Baggage.Current), metadata, MetadataGetter);
+                    if (propagationContext.ActivityContext.IsValid())
+                    {
+                        currentContext = propagationContext.ActivityContext;
+                        if (propagationContext.Baggage != default)
+                        {
+                            Baggage.Current = propagationContext.Baggage;
+                        }
+                    }

```

Comment:
```suggestion
                    if (propagationContext.ActivityContext.IsValid())
                    {
                        currentContext = propagationContext.ActivityContext;
                    }
                    if (propagationContext.Baggage != default)
                    {
                        Baggage.Current = propagationContext.Baggage;
                    }
```

Reply:
so propagation context is entirely separate from ActivityContext I guess?

Reply:
Ya that's what I was trying to get at. You could technically get Baggage by itself, although it probably doesn't make much sense to do that in practice.